src/read/tokio.rs

```diff
@@ -0,0 +1,328 @@
+        uncompressed_size: uncompressed_size as u64,
+        file_name,
+        file_name_raw: file_name_raw.into(),
+        extra_field: Some(Arc::new(extra_field)),
+        central_extra_field: None,
+        file_comment,
+        header_start: offset,
+        extra_data_start: None,
+        central_header_start,
+        data_start: OnceLock::new(),
+        external_attributes: external_file_attributes,
+        large_file: false,
+        aes_mode: None,
+        aes_extra_data_start: 0,
+        extra_fields: Vec::new(),
+    };
+
+    match parse_extra_field(&mut result).await {
+        Ok(..) | Err(ZipError::Io(..)) => {}
+        Err(err) => return Err(err),
+    }
+
+    let aes_enabled = result.compression_method == CompressionMethod::AES;
+    if aes_enabled && result.aes_mode.is_none() {
+        return Err(ZipError::InvalidArchive(
+            "AES encryption without AES extra data field",
+        ));
+    }
+
+    result.header_start = result
+        .header_start
+        .checked_add(archive_offset)
+        .ok_or(ZipError::InvalidArchive("Archive header is too large"))?;
+    Ok(result)
+}
+
+pub async fn parse_extra_field(file: &mut ZipFileData) -> ZipResult<()> {
+    let Some(extra_field) = &file.extra_field else {
+        return Ok(());
+    };
+    let mut reader = Cursor::new(extra_field.as_ref());
+
+    while (reader.position() as usize) < extra_field.len() {
+        let kind = reader.read_u16_le().await?;
+        let len = reader.read_u16_le().await?;
+        let mut len_left = len as i64;
+        match kind {
+            0x001 => {
+                if file.uncompressed_size == spec::ZIP64_BYTES_THR {
+                    file.large_file = true;
+                    file.uncompressed_size = reader.read_u64_le().await?;
+                    len_left -= 8;
+                }
+                if file.compressed_size == spec::ZIP64_BYTES_THR {
+                    file.large_file = true;
+                    file.compressed_size = reader.read_u64_le().await?;
+                    len_left -= 8;
+                }
+                if file.header_start == spec::ZIP64_BYTES_THR {
+                    file.header_start = reader.read_u64_le().await?;
+                    len_left -= 8;
+                }
+            }
+            0x9901 => {
+                if len != 7 {
+                    return Err(ZipError::UnsupportedArchive(
+                        "AES extra data field has an unsupported length",
+                    ));
+                    let vendor_version = reader.read_u16_le().await?;
+                    let vendor_id = reader.read_u16_le().await?;
+                    let mut out = [0u8];
+                    reader.read_exact(&mut out).await?;
+                    let aes_mode = out[0];
+                    #[allow(deprecated)]
+                    let compression_method =
+                        CompressionMethod::from_u16(reader.read_u16_le().await?);
+
```

Comment:
Factor out lines 286-313 into a shared method; it can be called `decode_aes_extra_data`.